proposals/Fleetctl-trigger.md

```diff
@@ -0,0 +1,95 @@
+  the scheduled jobs.
+- The trigger channel for each `schedule` is exposed via a new `schedules` map on the `Service` struct. 
+
+### New endpoint `GET /trigger?name={:name}` 
+- The request handler first calls `ds.Lock` to check if the named schedule is locked.
+- If the named schedule is unlocked, request handler sends a trigger signal on the schedule's
+  trigger channel and the server responds with status `202 Accepted`.  
+- If the named schedule is locked (presumably because the schedule is currently running), the server
+  responds with status `409 Conflict` and includes a message indicating the schedule is currently
+  locked. It is up to the user to retry. To facilitate retries, the response message could be expanded to
+  include additional status information, such as the expiration time of the current lock.
+
+### Schedule locks
+- Currently, lock duration is based on the schedule interval. 
+  - Once an instance takes the lock, it will hold the lock for the duration of the interval, even
+    after it has completed the jobs in the schedule. 
+  - For long-running jobs, the lock may expire before the current instance completes its run,
+    meaning that it is currently possible for another instance to start an overlapping job.
+  - If the lockholder instance is terminated or killed, locks are not released, which may frustrate
+    a user's attempt to configure a shorter schedule interval before the lock held by the dead
+    instance expires.
+
+- Under this proposal, locks become more dynamic.
+  - Current lockholder releases its lock once it finishes running the schedule.
+  - Graceful shutdown process handles release of locks upon termination signals. Jobs are
+    preemptable and an interrupt function must be specified for each job, 
+    e.g., `schedule.New(...).WithJob("job_name", jobFn, interruptFn)`.
+  - As a fallback for cases that can't be handled via graceful shutdown (e.g., `SIGKILL`), the
+    expiration for a lock is initially set to a relatively short default duration (e.g., 5 minutes).
+    The expiration is then periodically extended by the current instance so long as scheduled jobs
+    are running.  If the current instance dies without graceful shutdown, the lock will only be held
+    by the dead instance for a short period.
+
+### Additional UX considerations 
+- What are some potential options that would be useful for the `fleetctl trigger` command?
+  - Request the current status of the the named schedule without triggering a new run.
+    For example, `--status` could provide the user with the running time of the schedule (this would
+    require that we expand the `locks` table to include additional timestamp information, such as
+    lock start time and lock release time).
+  - Other useful options?
+
+- What rules should determine when the interval ticker resets? Consider the following cases where
+  `s.scheduleInterval = 1*time.Hour`: 
```

Comment:
@chiiph, does this help to clarify the issue?

Reply:
Hm, not entirely. I understood from the description above that if you run `fleetctl trigger ...` and there's a lock, then it'll just return a 4xx.

Reply:
Yes, that is the intended behavior for a trigger request. This question is getting at when the next scheduled (i.e. non-triggered) run should happen.

Reply:
Oh! I understand now what you mean. Hm, I would just wait another interval since it already happened.

Reply:
Ok, I think you are saying that once one job finishes, the next scheduled job will run after one full interval has passed.  This would have the side-effect of shifting interval forward over time. That could become problematic, especially for resource-intensive jobs where you'd like the job to run in off-hours. 

As things stand today, there isn't really a clear way to orchestrate the exact time jobs run.  For example, the 24-hour vuln interval might be running at 9am each day assuming that was when the very first vuln lock occurred.

On the other hand, I suppose fleetctl trigger command could be used as a way to peg the interval to a particular time. For example, a user might create a script to trigger a daily job at 3am each day.

Reply:
> That could become problematic, especially for resource-intensive jobs where you'd like the job to run in off-hours.

Is this a future we want to move towards? As in, do we want to have a way to run jobs explicitly off-hours? It doesn't sound like a bad idea, but it's also the first time I hear about it. So I'm not sure if optimizing for that is the best way to approach this.

Reply:
> So I'm not sure if optimizing for that is the best way to approach this.

I don't know if it is premature optimization necessarily. It feels like a natural extension of some basic decisions we need to make about how the trigger and interval interact.

Reply:
I'd go with simpler for now. Assuming the following is simple: triggering "schedules" should not impact when scheduled "schedules" run :)

Reply:
@lucasmrod, I'm all for simplicity here :) 

But there are a couple potential cases that I think we should clarify:

1. What if a triggered job runs immediately prior to a scheduled tick? 
Consider: Schedule ticks at the top of every hour. At 01:59, an admin triggers a run that takes 59 seconds to run.
- Should the next scheduled run happen just 1 second later at 02:00?

2. What if a triggered job runs across a scheduled tick? Consider: Schedule ticks at the top of every hour. At 01:59, an admin triggers a run that takes 5 minutes to run.
- Should the 02:00 scheduled run be skipped?
- Should the next scheduled run happen at 03:00?

Reply:
> Should the next scheduled run happen just 1 second later at 02:00?

Yes. For this first version I believe this is a-ok. We have this issue currently; multiple Fleet instances that started at different times, may cause the schedule to run more often than the configured intervals.

> Should the 02:00 scheduled run be skipped?

Yes, because the scheduled run will call `ds.Lock` and will return `false, nil` and not run; it will then try running at 03:00.

Let me know if I'm making sense.

Reply:
FWIW I would also expect `fleetctl trigger` to not impact the normal scheduling of jobs. It's an extra, manual trigger that forces it to run now if possible.